File: cpp/traverser.cpp
#include "builtin.hpp"
#include "stat.hpp"
#include "os/path.hpp"
#include "os/__init__.hpp"
#include "sys.hpp"
#include "printer.hpp"
#include "globals.hpp"
#include "traverser.hpp"

namespace __traverser__ {

str *const_0;

using __printer__::print_update;

str *__name__;



void *traversal_callback(str *_, str *dirname, list<str *> *files) {
    /**
    Function called during `os.path.walk` directory traversal.
    
    `os.path.walk` is deprecated, but `os.walk` is not yet implemented in
    ShedSkin.
    
    @param dirname: The name of the directory currently being examined.
    @type  dirname: str
    @param files: The list of file names residing within the current directory.
    @type  files: list or iterable
    */
    list<str *>::for_in_loop __12;
    list<str *> *__9;
    str *filename, *fullname;
    __ss_int __11;
    __iter<str *> *__10;

    fullname = const_0;
    __globals__::breadth = (__globals__::breadth+1);
    if (___bool(files)) {

        FOR_IN(filename,files,9,11,12)
            fullname = ((dirname)->__add__(__os__::__path__::sep))->__add__(filename);
            __globals__::now_length = ___max(2, 0, __globals__::max_length, len(fullname));
            if ((__globals__::max_length<__globals__::now_length)) {
                __globals__::max_length = __globals__::now_length;
                __globals__::longest_file = fullname;
            }
        END_FOR

    }
    else {
        __globals__::now_length = ___max(2, 0, __globals__::max_length, len(dirname));
        if ((__globals__::max_length<__globals__::now_length)) {
            __globals__::max_length = __globals__::now_length;
            __globals__::longest_file = dirname;
        }
    }
    __globals__::now_depth = len(dirname->split(__os__::sep));
    __globals__::now_depth = ___max(2, 0, __globals__::max_depth, (__globals__::now_depth-1));
    if ((__globals__::max_length<__globals__::now_length)) {
        __globals__::max_length = __globals__::now_length;
        __globals__::longest_file = dirname;
    }
    if ((__globals__::max_depth<__globals__::now_depth)) {
        __globals__::max_depth = __globals__::now_depth;
        __globals__::deepest_path = dirname;
    }
    if (__globals__::runas_program) {
        print_update(__globals__::breadth, __globals__::max_length, __globals__::max_depth);
    }
    return NULL;
}

void __init() {
    const_0 = new str("");

    __name__ = new str("traverser");

}

} // module namespace

File: deep/traverser.py
"""Logic for traversing the filesystem."""

import os

from . import globals  # Needed for ShedSkin

from .printer import print_update


def traversal_callback(_, dirname, files):
    """
    Function called during `os.path.walk` directory traversal.

    `os.path.walk` is deprecated, but `os.walk` is not yet implemented in
    ShedSkin.

    @param dirname: The name of the directory currently being examined.
    @type dirname: str
    @param files: The list of file names residing within the current directory.
    @type files: list or iterable
    """
    fullname = ''

    # "Breadth" is the total number of directories that have been examined.
    globals.breadth += 1

    # "Length" is the longest path name encountered during the traversal.
    if files:
        for filename in files:
            fullname = dirname + os.path.sep + filename
            globals.now_length = max(globals.max_length, len(fullname))
            if globals.max_length < globals.now_length:
                globals.max_length = globals.now_length
                globals.longest_file = fullname
    else:  # No files in this directory; check the name of the directory itself
        globals.now_length = max(globals.max_length, len(dirname))
        if globals.max_length < globals.now_length:
            globals.max_length = globals.now_length
            globals.longest_file = dirname

    # "Depth" is the largest subdirectory chain encountered during traversal.
    globals.now_depth = len(dirname.split(os.sep))
    globals.now_depth = max(globals.max_depth, globals.now_depth - 1)

    if globals.max_depth < globals.now_depth:
        globals.max_depth = globals.now_depth
        globals.deepest_path = dirname

    if globals.runas_program:
        print_update(globals.breadth, globals.max_length, globals.max_depth)
